File: 141_linked_list_cycle/Solution.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

def create_list(nums: list) -> ListNode:
    head = ListNode(nums[0])
    tmp = head
    for num in nums[1:]:
        tmp.next = ListNode(num)
        tmp = tmp.next
    tmp.next = head

    return head

class Solution:
    def hasCycle(self, head: ListNode) -> bool:
        if not head: return False

        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast: return True

        return False


if __name__ == '__main__':
    head = create_list([1, 2, 3, 4, 5])
    print(Solution().hasCycle(head))

File: 141_linked_list_cycle/Solution.java
class ListNode {
    int val;
    ListNode next;
    ListNode(int val) {
        this.val = val;
    }
}

public class Solution {
    public static ListNode creatList(int[] nums) {
        ListNode head = new ListNode(nums[0]);
        ListNode tmp = head;
        for (int i = 1; i < nums.length; i++) {
            tmp.next = new ListNode(nums[i]);
            tmp = tmp.next;
        }
        tmp.next = head;

        return head;
    }

    public boolean hasCycle(ListNode head) {
        if (head == null) return false;

        ListNode slow = head, fast = head;
        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (fast == slow) return true;
        }

        return false;
    }

    public static void main(String[] args) {
        ListNode head = creatList(new int[]{1, 2, 3, 4, 5});
        System.out.println(new Solution().hasCycle(head));
    }
}
